### PR TITLE
Fix led_strip overlay

### PR DESCRIPTION
**Pull request by haslinghuis**
Fixes: #2359

### CODE REVIEW

**Comment by mikeller** [comment]
Can you please explain why all of this is required to fix what seems to be a relatively small bug?

**Comment by haslinghuis** [comment]
Still can't find the problem not loading colors, functions and overlays.

**Comment by TheIsotopes** [comment]
I've tested it and it seems to be working again.

**Comment by haslinghuis** [comment]
Weird. I took the previous version of led_strip.js and I have exact the same behavior. Functions (Modes, Arm State, Battery, RSSI, GPS, Ring) are not visible after `save` and `re-entry` of the tab?

**Comment by haslinghuis** [comment]
Found a problem that could be the root of the cause:
`const ActiveFunctions = $('select.functionSelect'.val();` is undefined somehow. It doesn't help to define it earlier.

After clicking a LED `activeFunctions` becomes available, that's why it's working when configuring LED's.

So this could be the cause not assigning the functions/directions at load time. After that we can validate the `mask` for actual saving and loading the LED's.

I have no clue why this is happening.

File: src/js/msp/MSPHelper.js [hunk 1]
```diff
@@ -2615,15 +2617,15 @@ MspHelper.prototype.sendLedStripConfig = function(onCompleteCallback) {
             mask |= (led.y << 0);
             mask |= (led.x << 4);
 
-            for (const functionLetterIndex of led.functions) {
+            for (const functionLetterIndex in led.functions) {
```
Comment (mikeller) [inline src/js/msp/MSPHelper.js:2620]: Do we want to include all non-array properties of `led.functions` in this? If not then this is wrong, and the line below should be changed.

File: src/js/msp/MSPHelper.js [hunk 2]
```diff
@@ -2615,15 +2617,15 @@ MspHelper.prototype.sendLedStripConfig = function(onCompleteCallback) {
...
                 const fnIndex = ledBaseFunctionLetters.indexOf(led.functions[functionLetterIndex]);
                 if (fnIndex >= 0) {
                     mask |= (fnIndex << 8);
                     break;
                 }
             }
 
-            for (const overlayLetterIndex of led.functions) {
+            for (const overlayLetterIndex in led.functions) {
```
Comment (mikeller) [inline src/js/msp/MSPHelper.js:2628]: Same here.

File: src/js/msp/MSPHelper.js [hunk 3]
```diff
@@ -2632,7 +2634,7 @@ MspHelper.prototype.sendLedStripConfig = function(onCompleteCallback) {
 
             mask |= (led.color << 18);
 
-            for (const directionLetterIndex of led.directions) {
+            for (const directionLetterIndex in led.directions) {
```
Comment (mikeller) [inline src/js/msp/MSPHelper.js:2637]: Same here.

**Review by mikeller — CHANGES_REQUESTED** [review]

File: src/js/msp/MSPHelper.js (hunk 1 above)
Comment (haslinghuis) [reply src/js/msp/MSPHelper.js]: Changed back to standard for loop is much more clear.

Comment (mikeller) [reply src/js/msp/MSPHelper.js]: Using `for ( of )` would be perfectly fine in these cases:

```
            for (const directionLetter of led.directions) {
￼                const bitIndex = ledDirectionLetters.indexOf(directionLetter);
￼                if (bitIndex >= 0) {
￼                    directionMask = bit_set(directionMask, bitIndex);
                }
            }
```

However, `for ( in )` is not ok, as it iterates over a different set.

**Review by mikeller — APPROVED** [review]

File: src/js/msp/MSPHelper.js (hunk 1 above)
Comment (haslinghuis) [reply src/js/msp/MSPHelper.js]: Somehow for ( of ) didn't work here, and for ( in ) did, but it's recommended not to use it as index while iterating over an array:
https://stackoverflow.com/questions/500504/why-is-using-for-in-for-array-iteration-a-bad-idea

**Pull request merged**